packages/frontend-dynamic-feature-loader/src/loader.ts

```diff
@@ -0,0 +1,196 @@
+
+/**
+ *
+ * @public
+ */
+export type DynamicFrontendFeaturesLoaderOptions = {
+  /**
+   * Additional module federation arguments for the Module Federation runtime initialization.
+   */
+  moduleFederation: Omit<Parameters<typeof init>[0], 'name' | 'remotes'>;
+};
+
+/**
+ * A function providing a loader of frontend features exposed as module federation remotes
+ * from the backend dynamic features service.
+ *
+ * @public
+ */
+export function dynamicFrontendFeaturesLoader(
+  options?: DynamicFrontendFeaturesLoaderOptions,
+): CreateAppFeatureLoader {
+  return {
+    getLoaderName() {
+      return 'dynamic-plugins-loader';
+    },
+
+    async load({ config }) {
+      const dynamicPLuginsConfig = config.getOptionalConfig('dynamicPlugins');
+      if (!dynamicPLuginsConfig) {
+        return {
+          features: [],
+        };
+      }
+
+      function error(message: string, err: unknown) {
+        // eslint-disable-next-line no-console
+        console.error(
+          `${message}: ${
+            err instanceof Error ? err.toString() : JSON.stringify(err)
+          }`,
+        );
+      }
+
+      const appPackageName =
+        config.getOptionalString('app.packageName') ?? 'app';
+      let frontendPluginManifests: {
+        [key: string]: string;
+      };
+      try {
+        const apiClient = new DefaultApiClient({
+          discoveryApi: FrontendHostDiscovery.fromConfig(config),
+          fetchApi: {
+            fetch(input) {
+              return global.fetch(input);
+            },
+          },
+        });
+
+        const response = await apiClient.getManifests({});
+        if (!response.ok) {
+          throw new Error(`${response.status} - ${response.statusText}`);
+        }
+        frontendPluginManifests = await response.json();
+        if (typeof frontendPluginManifests !== 'object') {
+          throw new Error(`Invalid Json content: should be a Json object`);
+        }
+      } catch (err) {
+        error(
+          `Failed fetching module federation configuration of dynamic frontend plugins`,
+          err,
+        );
+        return { features: [] };
+      }
+
+      try {
+        init({
+          ...options?.moduleFederation,
+          name: appPackageName
+            .replaceAll('@', '')
+            .replaceAll('/', '__')
+            .replaceAll('-', '_'),
+          remotes: Object.entries(frontendPluginManifests).map(
+            ([name, manifestLocation]) => ({
+              name: name,
+              entry: manifestLocation,
+            }),
+          ),
+        });
+      } catch (err) {
+        error(`Failed initializing module federation`, err);
+        return { features: [] };
+      }
+
+      const features = (
+        await Promise.all(
+          Object.entries(frontendPluginManifests).map(
+            async ([name, manifestLocation]) => {
+              // eslint-disable-next-line no-console
+              console.info(
```

Comment:
i'll assume these are meant to be kind of temporary, but would it be better to only log once complete or in a `catch` if it failed? Saying "i'm about to ..." is less useful when we assume that they normally always succeed

Reply:
I changed them to debug calls, and kept the log level for the message related to load completion.

Reply:
Fixed in commit 4335b04da0394e2eb126146392ed527c301580a1